### ISSUE TITLE
Rewrite WebM support to work with our new MSE code

### ISSUE THREAD

**Issue by trav90** [MoonchildProductions/Pale-Moon#1119]
Dependent on #1033 and splitting from the discussion in #1116 since I think this is big enough to deserve its own issue.

The new code being implemented in #1033 only works with MP4 at this time. By flipping the media.mediasource.format-reader pref to enable our new MSE architecture effectively makes using MSE + WebM impossible (even if the box is checked in Preferences -> Content).

In order to use WebM + MSE, the above mentioned pref will have to be set back to false (which causes the user to miss out on the majority of MSE + MP4 fixes that I have been working on), so it is a bit of a catch 22.

But considering where we are right now and the fact I don't want to leave our MP4 code "good enough" in order to work on WebM, I'm going to finish whipping our MP4 code into shape for 27.4, and WebM can come in the next version.

Here's a list of things that need done before shipping this:

- [ ] Add a new WebMDemuxer object that will work together with the MediaFormatReader (this is similar to the MediaSourceDemuxer and MP4Demuxer and will be behind a pref).
- [ ] Update the MSE ContainerParser to work with the new WebMDemuxer.
- [ ] Update the TrackBuffersManager to fix a few WebM-specific issues with the new MSE.
- [ ] Enable all our WebM code by default.

**Comment by wolfbeast** [comment]
Should be simple enough to make the prefs linked from the UI, right? i.e.: if people enable WebM, flip the format reader pref if set, and v.v. -- should be all that's needed. (and people setting this manually in about:config are in the realm of draconian repercussions for doing stuff wrong, so... ;) )

**Comment by trav90** [comment]
I thought about that, but then on the other hand by disabling the new format reader to allow MSE + WebM they will experience regressions with MSE + MP4 (such as broken Twitch).

I did a little research today, and it appears that the vast majority of the new MSE code I have landed for use with MP4 (such as the TrackBufferManager among many other areas) will also be usable by WebM as-is or with minimal tweaks. The big thing we will need is a rewite of our WebM demuxer to work with the MediaFormatReader (much like MP4 and MSE); i.e. Much of the work I've already done lays some groundwork (and has given me experience doing this to boot) and will make adding working WebM + new MSE support easier.

So if I can get our MSE + MP4 code fixed up and flip the pref on for unstable builds for a while and get some community feedback on it, I may be able to get WebM fixed up enough to ship as well (depending on what the timetable is for 27.4).

**Comment by wolfbeast** [comment]
Yeah, you're right. If you can't get WebM in shape in time, it's probably best to make a temporary hard check in the media parsing module so that it will always use the old code for WebM regardless of the format-reader pref.

pseudo-code for what I mean:
```` C
if (media == WebM || formatreader-pref == false) {
  < use old path >
} else {
  < use new path >
}
````
I just hope that it'll be straightforward to check for media type at that location in the code.

**Comment by trav90** [comment]
Sounds good. I haven't really looked at the WebM parsing code yet, but hopefully it will be pretty straightforward to do. What I'm hoping to do is be able to find the source of and fix the shutdown hang quickly (having trouble reproducing it on Linux), and then spend the rest of this release cycle at least getting a base implementation of the new WebM in (which shouldn't take long). It wouldn't be as well tested as our current MP4 support, would be preffed off by default and not recommended for use yet, but that pretty well describes our current WebM code too.

**Comment by wolfbeast** [comment]
Looks like Vista's WMF isn't up to shape for MSE use https://forum.palemoon.org/viewtopic.php?f=3&t=16016&p=116408#p116401 which will make putting some form of temp check in something higher priority since they will have to fall back to WebM. Since the shutdown hang was resolved (and turned out unrelated to changing paths here) there's nothing blocking putting automatic flipping here, I think.

**Comment by trav90** [comment]
With PR #1209 incoming, I'm going to start working on getting a temp workaround for WebM over the next couple days.

**Comment by wolfbeast** [comment]
I don't think it should take a couple days to create this (otherwise it'd be too complex a solution). If that turns out to be the case, a front-end temporary workaround (flip off new format reader when people enable MSE+WebM and v.v.) is always also possible. The only people this will impact will be people on Vista and people who want WebM for 4k on YT (something Firefox doesn't even do, afaik).

Main issue I think will depend on whether you can control this on a per-media basis. If not, it'll likely be too complicated to weave in, but I'll let you judge.

**Comment by trav90** [comment]
I'll be working on this overnight and should have a status update in the morning.

**Comment by trav90** [comment]
OK, I think controlling this on a per-media basis is going to be too complex and difficult to do. There's not really an ideal solution here, but I think that the simplest choice for 27.4 will be the front-end temp workaround. In addition to the new option to enable/disable the updated MSE code (enabled by default), the MSE + WebM box needs to be "greyed out" or otherwise un-selectable as long as it is enabled (i.e. the new MSE code must be disabled before the MSE + WebM option is available).

I'm not to familiar with the code for the front-end options, so hopefully this will be doable without too much effort. Also, I think this should be restricted to the 27.4_RelBranch, leaving the unstable builds as-is in this regard (will make WebM testing easier once I start working on this).

**Comment by wolfbeast** [comment]
I can do the front-end pref juggling bit. I have plenty of experience with that code re: Pale Moon Commander.

>Also, I think this should be restricted to the 27.4_RelBranch, leaving the unstable builds as-is in this regard (will make WebM testing easier once I start working on this).

Indeed, which is one of the reasons why I split off that branch immediately after your last PR.
Your added option is a good idea; I'll just have to make sure that the string changes are going to be adopted by language packs before release, so it'll be some short-term wrangling that JustOff will have to do with the translators. I'll make a separate issue for this FE temporary code.

**Comment by trav90** [comment]
Sounds good! :+1: Thanks!

**Comment by trav90** [comment]
Most of the prep work for this has completed. There is still one major change Mozilla made that I need to research if it's more feasible or desirable to work around it or go ahead and port it to our tree. I've added a checklist to the first comment here for tracking.

**Comment by trav90** [comment]
Status update on this: The major change I mentioned above that Mozilla's WebM support is dependent on is a can of worms that I'd rather not open right now (it introduced several serious bugs/regressions, and at least one sec-critical crash in Firefox). So rather than porting that mess and fixing all the fallout, I'm going to attempt reverse-engineering Mozilla's WebM patches to work with our code in it's current state. Will be working on it in the coming week.

**Comment by trav90** [comment]
OK, after spending several days researching this and working with the code it seems that the Mozilla changes I have been attempting to work around are going to be needed after all. At a glance, the changes seemed to be about adapting more of the media subsystem to use the state machinery to avoid race conditions/increase stability. While that is part of the reasoning for the changes, the other part (which I've discovered through deeper research and testing) is to account for differences between WebM and MP4 (i.e., if we want WebM to work with the new MSE code then I will need to port the state machinery patches as well).

In the bigger picture, it'd probably be good to hookup more of our code to the state machinery anyway, as it will help to increase overall stability of the media subsystem and prevent race conditions. I also feel it might help with some of the various odd reports on the forum of media issues (much like #1209 resolved issues with seeking and live streams).

That said, the downsides are it's going to take a bit longer to get this fixed up, it will have a decent chance of introducing regressions, and it will break our WebM support in its current form (i.e. the new code will have to be at least good enough to enable it before we ship it).

**Comment by wolfbeast** [comment]
Clearing the milestone following e-mail discussion with Travis.

**Comment by trav90** [comment]
Considering the complexity and amount of time that will be required to complete this I'm considering this solved by UXP.
